src/algebra/algebra/spectrum.lean

```diff
@@ -0,0 +1,187 @@
+by { apply not_iff_not.mp, simp [set.not_not_mem, mem_spectrum_iff_not_unit] }
+
+lemma mem_resolvent_of_left_right_inverse {r : R} {a b c : A}
+  (h₁ : (r • 1 - a) * b = 1) (h₂ : c * (r • 1 - a) = 1) :
+  r ∈ resolvent R a :=
+units.is_unit ⟨r•1 - a, b, h₁, by rwa ←left_inv_eq_right_inv h₂ h₁⟩
+
+lemma mem_resolvent_iff {r : R} {a : A} :
+  r ∈ resolvent R a ↔ is_unit (r•1 - a) :=
+by { tidy }
+
+-- products of scalar units and algebra units
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, and units `r : units R`
+and `a : units A`, then `unit_mul_unit r a` constructs a `units A` with value `r•a`. -/
+
+lemma is_unit.smul_smul_sub_smul_iff {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - r • a) ↔ is_unit (s • 1 - a) :=
+begin
+  split,
+  { intro h',
+    have inv_smul_eq : r⁻¹•(r•s•1 - r•a) = s•1 - a,
+      by simp [smul_sub, smul_smul],
+    rw ←inv_smul_eq,
+    exact (r⁻¹ • h'.unit).is_unit, },
+  { intro h',
+    rw ←smul_sub,
+    exact (r • h'.unit).is_unit, },
+end
+
+lemma is_unit.smul_smul_sub_iff_is_unit_smul_sub_smul {r : units R} {s : R} {a : A} :
+  is_unit (r • s • 1 - a) ↔ is_unit (s • 1 - r⁻¹ • a) :=
+by { have h_eq : r•s•1 - r•(r⁻¹•a) = r•s•1 - a, by simp,
+     rw [←h_eq,is_unit.smul_smul_sub_smul_iff], }
+
+lemma is_unit.smul_sub_iff_is_unit_sub_smul {r : units R} {a : A} :
+  is_unit (r • 1 - a) ↔ is_unit (1 - r⁻¹ • a) :=
+begin
+  have with_smul_one : is_unit (r•(1 : R)•1 - a) ↔ is_unit ((1 : R)•1 - r⁻¹•a),
+    by exact is_unit.smul_smul_sub_iff_is_unit_smul_sub_smul,
+  simpa using with_smul_one,
+end
+
+lemma add_mem_spectrum {a : A} {r s : R} :
+  r ∈ σ a ↔ r + s ∈ σ (s • 1 + a) :=
+begin
+  apply not_iff_not.mpr,
+  simp only [mem_resolvent_iff],
+  have h_eq : (r+s)•1 - (s•1 + a) = r•1 - a,
+    by { rw add_smul, noncomm_ring },
+  simp [h_eq],
+end
+
+lemma mul_mem_spectrum_mul {a : A} {s : R} {r : units R} :
+  r • s ∈ σ (r • a) ↔ s ∈ σ a :=
+begin
+  apply not_iff_not.mpr,
+  simp only [mem_resolvent_iff],
+  have h_eq : (r•s)•(1 : A) = r•s•1, by simp,
+  rw h_eq,
+  exact is_unit.smul_smul_sub_smul_iff,
+end
+
+theorem left_add_coset_spectrum (a : A) (r : R) :
+  left_add_coset r (σ a) = σ (r • 1 + a) :=
+by { ext, rw [mem_left_add_coset_iff, neg_add_eq_sub, add_mem_spectrum],
+     nth_rewrite 1 ←sub_add_cancel x r, }
+
+-- this one isn't quite as simple because `R` is not a `group` (but it is an `add_group`)
+-- therefore we don't have access to `mem_left_add_coset_iff`
+theorem unit_left_coset_spectrum (a : A) (r : units R) :
+  left_coset ↑r (σ a) = σ (r • a) :=
+begin
+  ext,
+  split,
+  { rintro ⟨y,y_mem,hy⟩,
+    rw ←hy,
+    exact mul_mem_spectrum_mul.mpr y_mem, },
+  { intro hx,
+    have self_inv : x = r•r⁻¹•x, by simp,
+    rw self_inv at *,
+    exact mem_left_coset ↑r (mul_mem_spectrum_mul.mp hx), },
+end
+
+-- `r ∈ σ(a*b) ↔ r ∈ σ(b*a)` for any `r : units R`
+theorem unit_mem_spectrum_mul_iff_swap_mul {a b : A} {r : units R} :
+  ↑r ∈ σ (a * b) ↔ ↑r ∈ σ (b * a) :=
```

Comment:
I think this might be better as
```suggestion
  σ (a * b) = σ (b * a) :=
```

Reply:
Oh nevermind, that's false! I meant
```suggestion
  (coe : units R → R) ⁻¹' σ (a * b) = coe ⁻¹'  σ (b * a) :=
```
In which case it probably makes sense to keep both versions